lib/auth/default_current_user_provider.rb

```diff
@@ -240,14 +233,20 @@ def log_on_user(user, session, cookies, opts = {})
   end
 
   def cookie_hash(unhashed_auth_token, user)
-    cookie = DiscourseAuthCookie.new(
+    data = {
       token: unhashed_auth_token,
       user_id: user.id,
       trust_level: user.trust_level,
-      issued_at: Time.zone.now
-    )
+      issued_at: Time.zone.now.to_i
+    }
+    # create a temporary request that we can use to do the cookie encryption for us.
+    # don't pass the real @env to prevent potential pollution/side effects
+    temp_request = ActionDispatch::Request.new(Rails.application.env_config.dup)
+    temp_request.cookie_jar.encrypted[TOKEN_COOKIE] = { value: data }
```

Comment:
Did you find any pollution/side effects? We create `ActionDispatch::Request` objects elsewhere. [e.g. here](https://github.com/discourse/discourse/blob/520a0cfa2b2234bd683d79dc73bdac4bf227976b/lib/route_matcher.rb#L28-L28).

Also, it's quite important that some env variables (e.g. the cookie salts) are made available to the Request object. I see you used `Rails.application.env_config`, which provides the defaults. But what if some early middleware was customizing these values per-request?

Could we instead do something like this:
```
request = ActionDispatch::Request.new(env)
request.cookie_jar.encrypted[TOKEN_COOKIE] = {
  value: data,
  httponly: true,
  secure: SiteSetting.force_https
}

Reply:
Fixed in https://github.com/discourse/discourse/pull/14706/commits/4a30b475af509942b879985763e46a6f24206db3.